### ISSUE TITLE
Add blinded_routes param to probeForRoute, getRouteToDestination, payViaRoutes, etc.

### ISSUE THREAD

**Issue by adambor** [alexbosworth/lightning#192]
The 0.18.3-beta release of LND added the support for [bLIP-39](https://github.com/lightning/blips/blob/master/blip-0039.md) blinded paths. To support this essentially means the lib needs to have one more optional parameter for the calls to the above mentioned methods. A great plus would be to also allow parsing blinded paths from the BOLT11 payment request via the parsePaymentRequest.

**Comment by alexbosworth** [comment]
Yeah this is needed, although I am also not fully sure that it's actually possible

**Comment by alexbosworth** [comment]
The way that the encrypted routes works is that the receiver forms the onion packet and the payer just gets a blob

But a probe requires that you change the hash of the payment to a random hash

The problem is that the hash that you are going to need to change is inside that blob that you can't edit

**Comment by adambor** [comment]
The payment hash isn't part of that encrypted blob, which makes sense because it is anyway shared by all the hops on the route, so wouldn't make much sense to hide it.

**Comment by alexbosworth** [comment]
Oh yeah? I didn't look at it carefully so that's why I wasn't sure, but it seemed like it would be

If not then it seems pretty feasible sure

**Comment by alexbosworth** [comment]
OK I thought of another potential problem

I'm not sure what the errors look like, but we want an error with authentication from the destination, since we don't know the destination, how does that work?

**Comment by adambor** [comment]
I think that should still be possible, the node can just use a blinded public key (as specified in the blinded route) and sign with that.

**Comment by adambor** [comment]
Ahh okay, so that means no probing and for routing it only makes sense to get the route to the introductory node, so no change there needed. So what would make sense is just adding blinded paths to payViaRoutes and adjusting parsePaymentRequest to be able to parse that blinded path from a bolt11 invoice.

**Comment by alexbosworth** [comment]
Yeah I think those are as much as we can do, but I would like to see another solution because probing is useful

**Comment by alexbosworth** [comment]
As I mentioned I didn't really go into this too deeply myself but there is a discussion here you could read or contribute to if you are interested in this usage flow https://github.com/lightning/bolts/issues/1211

**Comment by adambor** [comment]
Yep, will check it out, in the meantime I went without probing and I am just checking if there is a route to the introduction nodes. To get the blinded routes from bolt11 payment request I've forked the bolt11 npm package and added bLIP-39 blinded route parsing: https://github.com/adambor/bolt11. Some notable findings:
1. You need one `getRouteToDestination()` call to each introduction node (running them in parallel and then using Promise.all makes sense here), where the destination is introduction node's public key - introduction node is the `blinded_node_pubkey` of the first `blinded_hop`.
2. You need to use the blinded path's `cltv_expiry_delta` as the `cltv_delta` param, not the CLTV delta of the invoice.
3. The amount you specify needs to include the fee of the blinded route, so you need to take: `fee_base_msat + pay_amount_msat * fee_proportional_millionths/1000000`
4. The payment secret is not included in the bolt11 payment request (it is part of the `blinded_hop` data), so you cannot specify it in the routing request.
5. Invoices with blinded hops tend to have more total hops, therefore higher `max_timeout_height` might be required, and you should also not call `getRouteToDestination()` for the blinded paths which have `cltv_expiry_delta` higher than your `max_timeout_height` delta - this results in error being thrown.

**Comment by alexbosworth** [comment]
I think decodePaymentRequest is probably the first thing, it currently doesn't parse the BlindedPaymentPath elements

Then the route stuff can hopefully take that as an argument

And parsePaymentRequest doesn't have an lnd dependency so it has to be parsing the data from scratch

All on my todo list
